docs: add fraud proofs documentation (#160)
celestiaorg/quantum-gravity-bridge: docs/fraud-proofs.md

# QGB fraud proofs

> A Celestium is an L2 rollup that uses Celestia for data availability but settles on Ethereum or any other EVM L1 chain. In simpler terms, it's a layer 2 solution that leverages the Celestia blockchain security to ensure that data is available, while the actual transactions and smart contract interactions are settled on the Ethereum blockchain or any other EVM-compatible layer 1 blockchain.

Reviewer: Technically could be a layer 3 on a layer 2!

```suggestion
A Celestium is an rollup that uses Celestia for data availability but settles on Ethereum or any other EVM chain. In simpler terms, it's a layer 2 solution that leverages the Celestia blockchain security to ensure that data is available, while the actual transactions and smart contract interactions are settled on the Ethereum blockchain or any other EVM-compatible blockchain.
```

> Celestiums use the Quantum-Gravity-Bridge (QGB) contract, deployed on an EVM chain, to verify that its transaction data has been published to Celestia correctly. And would rely on fraud proofs otherwise.

Reviewer: ```suggestion
Celestiums use the Quantum Gravity Bridge (QGB) contract, deployed on an EVM chain, to verify that their transaction data has been published to Celestia correctly. Alternatively, they can use an optimistic scheme, only verifying transaction data inclusion during fraud proof execution.
```

# Fraud proofs

> Fraud proofs is the mechanism used to inform light clients in the case of an invalid rollup state transition or unavailable rollup block data. They rely on rollup full nodes getting the data that was published to Celestia, and executing all the state transitions to verify the rollup state. If they discover an invalid state transition or unavailable rollup data, they emit a fraud proof with the necessary information to convince light clients that fraud happened. This allows for trust minimized light clients, as the network only needs one honest full node to create the fraud proof and propagate it.

Reviewer: ```suggestion
Fraud proofs is the mechanism used to inform light clients in the case of an invalid rollup state transition or unavailable rollup block data. They rely on rollup full nodes getting the data that was published to Celestia, and executing all the state transitions to verify the rollup state. If they discover an invalid state transition or unavailable rollup data, they emit a fraud proof with the necessary information to convince light clients that fraud happened. This allows for trust-minimized light clients, as the network only needs one honest full node to create the fraud proof and propagate it.
```

If the Celestium is settlement smart contract based, then the contract would only need to receive a fraud proof to decide whether data was published correctly or not.

## Rollup header

Rollups can adopt many approaches to prove that fraud happened. One of which could be having the following fields in the rollup header:

- Rollup block state root
- A sequence of spans in Celestia: which references where the rollup data was published in the Celestia chain.

> Note: The sequence of spans can be defined using the following: `Height`, `start index`, and `length` in the Celestia block, in the case of a single Celestia block. However, it could be generalized to span over multiple blocks.

Reviewer: https://github.com/orgs/community/discussions/16925

```suggestion
> [!NOTE]  
> The sequence of spans can be defined using the following: `Height`, `start index`, and `length` in the Celestia block, in the case of a single Celestia block. However, it could be generalized to span over multiple blocks.
```

For the rest of the document, we will suppose that the sequence of spans only references one Celestia block.

## Unavailable data fraud proof

By construction, the rollup block data is the sequence of spans defined in the header. Thus, proving that the data is unavailable goes back to proving that the sequence of spans doesn't belong to the Celestia block, i.e. the span is out of bounds.

> We could prove that via creating a binary [merkle proof](https://github.com/celestiaorg/celestia-core/blob/c3ab251659f6fe0f36d10e0dbd14c29a78a85352/crypto/merkle/proof.go#L19-L31) of any row/column to the Celestia data root. This proof will provide the `Total` which is the number of rows/columns in the extended data square. And, we can use that to calculate the square size.

Reviewer: ```suggestion
We could prove that via creating a binary [Merkle proof](https://github.com/celestiaorg/celestia-core/blob/c3ab251659f6fe0f36d10e0dbd14c29a78a85352/crypto/merkle/proof.go#L19-L31) of any row/column to the Celestia data root. This proof will provide the `Total` which is the number of rows/columns in the extended data square. And, we can use that to calculate the square size.
```

Then, we will use that information to check if the provided transaction index, in the header, is out of the square size bounds.

> For the data root, we will use a binary merkle proof to prove its inclusion in a data root tuple root that was committed to by the QGB smart contract. More on this in [here](#1-data-root-inclusion-proof).

Reviewer: ```suggestion
For the data root, we will use a binary Merkle proof to prove its inclusion in a data root tuple root that was committed to by the QGB smart contract. More on this in [here](#1-data-root-inclusion-proof).
```

## Invalid transaction fraud proof

In order to prove an invalid transaction in the rollup, we need to prove the following:

- Prove that that transaction was posted to Celestia
- Prove that the transaction is invalid: left to the rollup to define.

The first part, proving that the transaction was posted to Celestia, can be done in three steps:

1. Prove that the data root is committed to by the QGB smart contract
2. Inclusion proof of the transaction to Celestia data root
3. Prove that that transaction is in the rollup sequence spans

### 1. Data root inclusion proof

> To prove the data root is committed to by the QGB smart contract, we will need to provide a merkle proof of the data root tuple to a data root tuple root. This can be created using the [`data_root_inclusion_proof`](https://github.com/celestiaorg/celestia-core/blob/c3ab251659f6fe0f36d10e0dbd14c29a78a85352/rpc/client/http/http.go#L492-L511) query.

Reviewer: ```suggestion
To prove the data root is committed to by the QGB smart contract, we will need to provide a Merkle proof of the data root tuple to a data root tuple root. This can be created using the [`data_root_inclusion_proof`](https://github.com/celestiaorg/celestia-core/blob/c3ab251659f6fe0f36d10e0dbd14c29a78a85352/rpc/client/http/http.go#L492-L511) query.
```

### 2. Transaction inclusion proof

> To prove that a rollup transaction is part of the data root, we will need to provide two proofs: a namespace merkle proof of the transaction to a row root. This could be done via proving the shares that contain the transaction to the row root using a namespace merkle proof. And, a binary merkle proof of the row root to the data root.

Reviewer: ```suggestion
To prove that a rollup transaction is part of the data root, we will need to provide two proofs: a namespace Merkle proof of the transaction to a row root. This could be done via proving the shares that contain the transaction to the row root using a namespace Merkle proof. And, a binary Merkle proof of the row root to the data root.
```

These proofs can be generated using the [`ProveShares`](https://github.com/celestiaorg/celestia-core/blob/c3ab251659f6fe0f36d10e0dbd14c29a78a85352/rpc/client/http/http.go#L526-L543) query.

### 3. Transaction part of the rollup sequence

The Celestia block is a 2d matrix of rows and columns, where the row roots and column roots are committed to in the data root. We can use this property to point which rows and columns does the transaction shares belong to. Then, we will use those coordinates to calculate the row major index of the transaction, and verify if it is part of the sequence span.

> The coordinates can be gotten using a namespace merkle proof of the shares to the row root, and also to the column root. Then, we will create merkle proofs of that row root and column root to the data root. These [proofs](https://github.com/celestiaorg/celestia-core/blob/c3ab251659f6fe0f36d10e0dbd14c29a78a85352/crypto/merkle/proof.go#L19-L31) will contain the [`Index`](https://github.com/celestiaorg/celestia-core/blob/c3ab251659f6fe0f36d10e0dbd14c29a78a85352/crypto/merkle/proof.go#L28) of the row/column, and the [`Total`](https://github.com/celestiaorg/celestia-core/blob/c3ab251659f6fe0f36d10e0dbd14c29a78a85352/crypto/merkle/proof.go#L27) number of rows/columns in the proofs which would allow us to calculate the row major index of the transaction.

Reviewer: ```suggestion
The coordinates can be gotten using a namespace Merkle proof of the shares to the row root, and also to the column root. Then, we will create Merkle proofs of that row root and column root to the data root. These [proofs](https://github.com/celestiaorg/celestia-core/blob/c3ab251659f6fe0f36d10e0dbd14c29a78a85352/crypto/merkle/proof.go#L19-L31) will contain the [`Index`](https://github.com/celestiaorg/celestia-core/blob/c3ab251659f6fe0f36d10e0dbd14c29a78a85352/crypto/merkle/proof.go#L28) of the row/column, and the [`Total`](https://github.com/celestiaorg/celestia-core/blob/c3ab251659f6fe0f36d10e0dbd14c29a78a85352/crypto/merkle/proof.go#L27) number of rows/columns in the proofs which would allow us to calculate the row major index of the transaction.
```

> # QGB fraud proofs

Reviewer 2: ```suggestion
# QGB Fraud Proofs
```

> A Celestium is a rollup that uses Celestia for data availability but settles on Ethereum or any other EVM chain. In simpler terms, it's a layer 2 solution that leverages the Celestia blockchain security to ensure that data is available, while the actual transactions and smart contract interactions are settled on the Ethereum blockchain or any other EVM-compatible blockchain.
> 
> Celestiums use the Quantum Gravity Bridge (QGB) contract, deployed on an EVM chain, to verify that their transaction data has been published to Celestia correctly. Alternatively, they can use an optimistic scheme, only verifying transaction data inclusion during fraud proof execution.

Reviewer 2: I think we can reduce this slightly by deduping a bit, while also conveying a rather complete intro by saying something along the lines of:

```suggestion
## Celestiums Intro

A Celestium is a rollup that uses Celestia for data availability but settles on any EVM chain. This works by the Celestia validator set periodically signing over batched data commitments and validator set updates, which are relayed to the QGB smart contract. The data commitments are stored in the EVM chain's state, and can be used to prove inclusion of any data posted to Celestia.
```

or we might not even need the intro at all, but I'll defer to the author

> # Fraud proofs

Reviewer 2: ```suggestion
# Fraud Proofs
```

Fraud proofs is the mechanism used to inform light clients in the case of an invalid rollup state transition or unavailable rollup block data. They rely on rollup full nodes getting the data that was published to Celestia, and executing all the state transitions to verify the rollup state. If they discover an invalid state transition or unavailable rollup data, they emit a fraud proof with the necessary information to convince light clients that fraud happened. This allows for trust-minimized light clients, as the network only needs one honest full node to create the fraud proof and propagate it.

> If the Celestium is settlement smart contract based, then the contract would only need to receive a fraud proof to decide whether data was published correctly or not.

Reviewer 2: are there types of celestiums that aren't smart contract based?

Author: I guess they all are smart contract based, but light clients also can verify the proofs before the settlement smart contract executes them.

## Rollup header

> Rollups can adopt many approaches to prove that fraud happened. One of which could be having the following fields in the rollup header:

Reviewer 2: ```suggestion
Rollups can adopt many approaches to prove that fraud occurred. One of which could be having the following fields in the rollup header:
```

- Rollup block state root

> - A sequence of spans in Celestia: which references where the rollup data was published in the Celestia chain.

Reviewer 2: > A sequence of spans

what is a span?

Author: something like this: https://github.com/celestiaorg/quantum-gravity-bridge/blob/28ffbdd1f20798b410ee2d3a213e70df8edfebd3/src/lib/verifier/test/RollupFraudProof.t.sol#L22-L35

> [!NOTE]  
> The sequence of spans can be defined using the following: `Height`, `start index`, and `length` in the Celestia block, in the case of a single Celestia block. However, it could be generalized to span over multiple blocks.

> For the rest of the document, we will suppose that the sequence of spans only references one Celestia block.

Reviewer 2: > sequence of spans

sequence of data?

Author: something like this: https://github.com/celestiaorg/quantum-gravity-bridge/blob/28ffbdd1f20798b410ee2d3a213e70df8edfebd3/src/lib/verifier/test/RollupFraudProof.t.sol#L22-L35

> For the rest of the document, we will suppose that the sequence of spans only references one Celestia block.

Reviewer 2: what occurs when the celestium needs to prove inclusion to multiple chunks of data?

Author: That depends on how they want to separate their data. Generally, they would have multiple references to data in Celestia blocks and they would prove inclusion to one of them

## Unavailable data fraud proof

By construction, the rollup block data is the sequence of spans defined in the header. Thus, proving that the data is unavailable goes back to proving that the sequence of spans doesn't belong to the Celestia block, i.e. the span is out of bounds.

We could prove that via creating a binary [Merkle proof](https://github.com/celestiaorg/celestia-core/blob/c3ab251659f6fe0f36d10e0dbd14c29a78a85352/crypto/merkle/proof.go#L19-L31) of any row/column to the Celestia data root. This proof will provide the `Total` which is the number of rows/columns in the extended data square. And, we can use that to calculate the square size.

Then, we will use that information to check if the provided transaction index, in the header, is out of the square size bounds.

> For the data root, we will use a binary Merkle proof to prove its inclusion in a data root tuple root that was committed to by the QGB smart contract. More on this in [here](#1-data-root-inclusion-proof).

Reviewer 2: this is important info! but I'm not sure making it its own type of fraud proof is our best option. Or imo we pick a different name for the proof.

The main reasoning I'm seeing not to is because we want fraud proofs to be as normal as possible. The only thing that changes here is if an error is thrown when checking for inclusion, then we know that the data was not posted to celestia. This is essentially all we need to communicate, that there is an edge case where the location in the header doesn't exist, in which case no further action is required to verify the normal fraud proof.

Author: This is just providing a distinction of the type of frauds that can happen from the QGB/Celestia perspective so that rollups can incorporate them in their proofs. 

What do you think about switching the titles:
`## Unavailable data fraud proof` => `## Proving unavailable data`?

### 2. Transaction inclusion proof

To prove that a rollup transaction is part of the data root, we will need to provide two proofs: a namespace Merkle proof of the transaction to a row root. This could be done via proving the shares that contain the transaction to the row root using a namespace Merkle proof. And, a binary Merkle proof of the row root to the data root.

These proofs can be generated using the [`ProveShares`](https://github.com/celestiaorg/celestia-core/blob/c3ab251659f6fe0f36d10e0dbd14c29a78a85352/rpc/client/http/http.go#L526-L543) query.

### 3. Transaction part of the rollup sequence

The Celestia block is a 2d matrix of rows and columns, where the row roots and column roots are committed to in the data root. We can use this property to point which rows and columns does the transaction shares belong to. Then, we will use those coordinates to calculate the row major index of the transaction, and verify if it is part of the sequence span.

> The coordinates can be gotten using a namespace Merkle proof of the shares to the row root, and also to the column root. Then, we will create Merkle proofs of that row root and column root to the data root. These [proofs](https://github.com/celestiaorg/celestia-core/blob/c3ab251659f6fe0f36d10e0dbd14c29a78a85352/crypto/merkle/proof.go#L19-L31) will contain the [`Index`](https://github.com/celestiaorg/celestia-core/blob/c3ab251659f6fe0f36d10e0dbd14c29a78a85352/crypto/merkle/proof.go#L28) of the row/column, and the [`Total`](https://github.com/celestiaorg/celestia-core/blob/c3ab251659f6fe0f36d10e0dbd14c29a78a85352/crypto/merkle/proof.go#L27) number of rows/columns in the proofs which would allow us to calculate the row major index of the transaction.

Reviewer 2: I'm not sure this is the best spot for this or if the user needs to know these things when thinking about fraud proofs. 

mostly, we just need to communicate what can possibly occur when using the fraud proofs. Either the inclusion proof is valid or invalid, and if its invalid what are the reasons for it being invalid (the portion about the data not being posted to the square).

I think this document is generally not making a clear enough distinction between what the qgb contracts are doing, which is just inclusion proofs, not really fraud proofs. 

Fraud proofs are important, but I think keeping those more separate is good.

## Proving unavailable data

By construction, the rollup block data **is the sequence of spans defined in the header**. Thus, proving that the data is unavailable goes back to proving that the sequence of spans doesn't belong to the Celestia block, i.e. the span is out of bounds.

We could prove that via creating a binary [Merkle proof](https://github.com/celestiaorg/celestia-core/blob/c3ab251659f6fe0f36d10e0dbd14c29a78a85352/crypto/merkle/proof.go#L19-L31) of any row/column to the Celestia data root. This proof will provide the `Total` which is the number of rows/columns in the extended data square. And, we can use that to calculate the square size.

Then, we will use that information to check if the provided transaction index, in the header, is out of the square size bounds.

For the data root, we will use a binary Merkle proof to prove its inclusion in a data root tuple root that was committed to by the QGB smart contract. More on this in [here](#1-data-root-inclusion-proof).

## Proving an invalid state transition

In order to prove an invalid transaction in the rollup, we need to prove the following:

> - Prove that that transaction was posted to Celestia
> - Prove that the transaction is invalid: left to the rollup to define.

Reviewer: ```suggestion
- Prove that the transaction was posted to Celestia, and
- Prove that the transaction is invalid. This is left to the rollup to define.
```

The first part, proving that the transaction was posted to Celestia, can be done in three steps:

> 1. Prove that the data root is committed to by the QGB smart contract

Reviewer: ```suggestion
1. Prove that the data root tuple is committed to by the QGB smart contract
```

> 2. Inclusion proof of the transaction to Celestia data root

Reviewer: ```suggestion
2. Verify inclusion proof of the transaction to Celestia data root
```

> 3. Prove that that transaction is in the rollup sequence spans

Reviewer: ```suggestion
3. Prove that the transaction is in the rollup sequence spans
```
